sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/checks/PropertyGetterAndSetterUsageCheck.kt

```diff
@@ -0,0 +1,147 @@
+import org.jetbrains.kotlin.psi.psiUtil.isPublic
+import org.jetbrains.kotlin.resolve.BindingContext
+import org.jetbrains.kotlin.types.KotlinType
+import org.jetbrains.kotlin.types.isNullable
+import org.sonar.check.Rule
+import org.sonar.check.RuleProperty
+import org.sonarsource.kotlin.api.AbstractCheck
+import org.sonarsource.kotlin.api.returnType
+import org.sonarsource.kotlin.api.secondaryOf
+import org.sonarsource.kotlin.plugin.KotlinFileContext
+
+private val GETTER_PREFIX = Regex("""^(get|is)\p{javaUpperCase}""")
+private val SETTER_PREFIX = Regex("""^set\p{javaUpperCase}""")
+
+@Rule(key = "S6512")
+class PropertyGetterAndSetterUsageCheck : AbstractCheck() {
+
+    @RuleProperty(
+        key = "checkPublicClasses",
+        description = "Check getters and setters on public classes",
+        defaultValue = "false"
+    )
+    var checkPublicClasses = false
+
+    override fun visitClass(klass: KtClass, ctx: KotlinFileContext) {
+        if (klass.isPublic && !checkPublicClasses) {
+            // avoid asking to change for property get() and set(value) to not break a public API
+            return
+        }
+        val classBody = klass.body ?: return
+        val javaAccessors = classBody.functions
+            .filter { isGetterOrSetter(it.valueParameters.size, it.name ?: "") }
+            .groupBy { it.name ?: "" }
+
+        if (javaAccessors.isEmpty()) {
+            return
+        }
+        val bindingCtx = ctx.bindingContext
+        classBody.properties
+            .filter { isPrivateWithoutGetterAndSetter(it) }
+            .forEach { checkProperty(it, javaAccessors, ctx, bindingCtx) }
+    }
+
+    private fun checkProperty(
+        prop: KtProperty, javaAccessors: Map<String, List<KtNamedFunction>>, ctx: KotlinFileContext, bindingCtx: BindingContext
+    ) {
+        prop.nameIdentifier?.let { propIdentifier ->
+            prop.typeReference?.let { bindingCtx.get(BindingContext.TYPE, it) }?.let { propType ->
+                val getterIdentifier = findJavaStyleGetterFuncIdentifier(propIdentifier.text, propType, javaAccessors, bindingCtx)
+                val setterIdentifier = findJavaStyleSetterFuncIdentifier(propIdentifier.text, propType, javaAccessors, bindingCtx)
+                if (getterIdentifier != null) {
+                    val secondaries = when (setterIdentifier) {
+                        null -> listOf(ctx.secondaryOf(propIdentifier, """Property "${propIdentifier.text}""""))
+                        else -> listOf(
+                            ctx.secondaryOf(propIdentifier, """Property "${propIdentifier.text}""""),
+                            ctx.secondaryOf(setterIdentifier, """Setter to convert to a "set(value)"""")
+                        )
+                    }
+                    ctx.reportIssue(
+                        getterIdentifier,
+                        """Convert this getter to a "get()" on the property "${propIdentifier.text}".""",
```

Comment:
If the messages have the same template, you can extract reporting part to a separate method